### PR TITLE
Improve rules for nameonly/positional/nameless and required/optional …

### LINKED ISSUE #3859: `nameonly` should not constrain default values
### Dafny version

3.13

### Code to produce this issue

```dafny
datatype Foo = Foo(
  nameonly bar: Option<string> := None,
  nameonly baz: Option<string>
)


This is probably because `Foo("positional parameters", "have order")`.
However `nameonly` means that I MUST

```dafny
Foo(
  baz := "Look ma, no order!"
)
```
```


### Command to run and resulting output

_No response_

### What happened?

In the above example `baz` gets:
`a required parameter must precede all optional parameters`

### What type of operating system are you experiencing the problem on?

Mac

### PR DESCRIPTION
**Pull request by RustanLeino**
Fixes #3859

This PR allows more parameter declarations (for example, as suggested in #3859, a `nameonly` parameter with a default value can now be follow by other `nameonly` parameters), clarifies some error messages, and gives new errors for useless situations (for example, a parameter's default value can not be used if it is followed by a positional-only parameter).

## The new rules

Formal parameters have three ways to indicate how they are to be passed in:

* `nameonly`: the only way to give a specific argument value is to name the parameter
* positional only: these are nameless parameters (which are allowed only for datatype constructor parameters)
* either positional or by name: this is the most common parameter

A parameter is either required or optional:
* required: a caller has to supply an argument
* optional: the parameter has a default value that is used if a caller omits passing a specific argument

The syntax for giving a positional-only (i.e., nameless) parameter does not a default-value expression, so a positional-only parameter is always required.

At a call site, positional arguments are not allowed to follow named arguments. Therefore, if `x` is a nameonly parameter, then there is no way to supply the parameters after `x` by position. Thus, any parameter that follows `x` must either by passed by name or have a default value. That is, if a later parameter does not have a default value, it is _effectively_ nameonly. We impose the rule that

* an effectively nameonly parameter must be declared as nameonly

For a positional-only parameter `x`, every parameter preceding `x` is _effectively_ required. We impose the rule that

* an effectively required parameter must not have a default-value expression

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Review by seebees — DISMISSED** [review]
LGTM

File: Source/DafnyCore/Resolver/NameResolutionAndTypeInference.cs [hunk 1]
```diff
@@ -4673,26 +4673,53 @@ public XConstraint_EquatableArg(IToken tok, Type a, Type b, bool allowSuperSub,
...
 
+      // Formal parameters have three ways to indicate how they are to be passed in:
+      //   * nameonly: the only way to give a specific argument value is to name the parameter
+      //   * positional only: these are nameless parameters (which are allowed only for datatype constructor parameters)
+      //   * either positional or by name: this is the most common parameter
+      // A parameter is either required or optional:
+      //   * required: a caller has to supply an argument
+      //   * optional: the parameter has a default value that is used if a caller omits passing a specific argument
+      //
+      // The syntax for giving a positional-only (i.e., nameless) parameter does not a default-value expression, so
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Resolver/NameResolutionAndTypeInference.cs:4684]: `does not a` => `does not allow a`

File: Source/DafnyCore/Resolver/NameResolutionAndTypeInference.cs [hunk 2]
```diff
@@ -4673,26 +4673,53 @@ public XConstraint_EquatableArg(IToken tok, Type a, Type b, bool allowSuperSub,
...
+      // a positional-only parameter is always required.
+      //
+      // At a call site, positional arguments are not allowed to follow named arguments. Therefore, if "x" is
+      // a nameonly parameter, then there is no way to supply the parameters after "x" by position. Thus, any
+      // parameter that follows "x" must either by passed by name or have a default value. That is, if a later
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Resolver/NameResolutionAndTypeInference.cs:4689]: `be passed`

File: Source/DafnyCore/Resolver/NameResolutionAndTypeInference.cs [hunk 3]
```diff
@@ -4673,26 +4673,53 @@ public XConstraint_EquatableArg(IToken tok, Type a, Type b, bool allowSuperSub,
...
-          reporter.Error(MessageSource.Resolver, formal.tok, "a required parameter must precede all optional parameters");
-        }
-        if (!allowNamelessParameters && !formal.HasName) {
-          reporter.Error(MessageSource.Resolver, formal.tok, "a nameless parameter must precede all nameonly parameters");
+          previousParametersWithDefaultValue.Add(formal);
+        } else if (nameOfMostRecentNameonlyParameter != null && !formal.IsNameOnly) {
+          // "formal" is preceded by a nameonly parameter, but itself is neither nameonly nor has a default value
+          reporter.Error(MessageSource.Resolver, formal.tok,
+            $"this parameter is effectively nameonly (because of the earlier nameonly parameter '{nameOfMostRecentNameonlyParameter}'); " +
+            "declare it as nameonly or give it an default-value expression");
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Resolver/NameResolutionAndTypeInference.cs:4719]: `a default-value expression`

File: Test/dafny0/ParameterResolution.dfy [hunk 4]
```diff
@@ -275,15 +275,15 @@ module NameOnlyParameters {
     least predicate LP(a: int, nameonly b: int, c: int := 100)
     static greatest predicate GP(a: int, nameonly b: int := 75, c: int := 100)
   }
-  iterator Iter(nameonly u: int, x: int)
+  iterator Iter(nameonly u: int, x: int) // error (x2, because of the desugaring of iterators): x is effectively nameonly, but not declared as such
```
Comment (keyboardDrummer) [inline Test/dafny0/ParameterResolution.dfy:278]: Maybe we should come up with a testing format where you can specify the expected error messages inline : )

**Review by keyboardDrummer — DISMISSED** [review]

**Comment by keyboardDrummer** [comment]
Such a nice quality of life improvement, thanks !

File: Test/dafny0/ParameterResolution.dfy (hunk 4 above)
Comment (RustanLeino) [reply Test/dafny0/ParameterResolution.dfy:278]: I do like being able to read the expected outcome in the test file itself. Therefore, I always put such "error" comments in the tests. But these comments aren't the exact error messages, don't capture column information, and in some cases are even placed only on a nearby line. It's plausible that we could turn these into a formally recognized testing file format.

**Pull request merged**